src/iso.c

```diff
@@ -0,0 +1,489 @@
+/*
+Copyright (C) 2016 The University of Notre Dame
+This software is distributed under the GNU General Public License.
+See the file LICENSE for details.
+*/
+
+/*
+ * This file has external facing iso_f and iso_d commands, but also
+ * contains iso_media_ commands for open, read, and write. The idea is
+ * that iso_f and iso_d are in regards to files and directories, while
+ * iso_media_ treats the entire iso storage device as a single, long file
+ * without an EOF which we can open, close, and seek around. rather than
+ * doing absurd byte counting in each of the iso_f and iso_d commands.
+ */
+
+#include "iso.h"
+#include "memory.h"
+#include "string.h"
+#include "kerneltypes.h"
+#include "ata.h"
+#include "console.h"
+#include "keyboard.h"
+
+#define ISO_BLOCKSIZE 2048
+#define PVD_OFFSET 16 * ISO_BLOCKSIZE
+#define ROOT_DR_OFFSET (PVD_OFFSET) + 156
+
+#define SEEK_CUR 0
+#define SEEK_SET 1
+
+// Use these to prevent multiple fetches of the same
+// ATAPI block for consecutive, single character reads
+char global_atapi_block[ATAPI_BLOCKSIZE];
+int global_atapi_unit = -1;
+int global_atapi_extent = -1;
+
+struct iso_point {
+    int ata_unit;
+    int cur_extent;
+    int cur_offset;
+};
+
+int get_directory_record(struct iso_point *iso_p, struct directory_record *dr);
+int hex_to_int(char *src, int len);
+int is_dir(int flags);
+int is_valid_record(struct directory_record *dr);
+long int iso_look_up(const char *pname, int *dl, int ata_unit);
+long int iso_recursive_look_up (const char *pname, struct iso_point *iso_p, int *dl);
+void iso_media_close(struct iso_point *iso_p);
+struct iso_point *iso_media_open(int ata_unit);
+int iso_media_read(void *dest, int elem_size, int num_elem, struct iso_point *stream);
+void iso_media_seek(struct iso_point *iso_p, long offset, int whence);
+
+/**
+ * @brief Reads the next directory record in the extent
+ * @details Starting at the current iso_p, reads a directory record into
+ * the argument dr, moving iso_p forward.
+ *
+ * @param iso_p A pointer to a struct iso_point which is set to the beginning of a directory record in a directory extent
+ * @param dr A pointer to a directory record to be filled with the information
+ * @return 1 on success, 0 on failure
+ */
+int get_directory_record(struct iso_point *iso_p, struct directory_record *dr) {
+    int bytes_read = 0;
+    int max_dr_size = 64;
+    char dr_dest[max_dr_size];
+    int result = iso_media_read(dr_dest, 1, max_dr_size, iso_p);
+    if(result != max_dr_size) {
+        console_printf("get directory record fails\n");
+        return 0;
+    }
+
+    memcpy(&(dr->length_of_record), dr_dest + 0, sizeof(dr->length_of_record));   //1 for bytes read
+    memcpy(&(dr->length_of_ext_record), dr_dest + 1, sizeof(dr->length_of_ext_record));
```

Comment:
Lots of magic numbers in here... Idk if its reasonable to move them all into constants, but we should at least leave a comment with a link to the page that tells what they all are.